frame/tx-pause/src/lib.rs

```diff
@@ -0,0 +1,206 @@
+// limitations under the License.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use frame_support::{
+	pallet_prelude::*,
+	traits::{CallMetadata, Contains, GetCallMetadata},
+};
+use frame_system::pallet_prelude::*;
+use sp_std::{convert::TryInto, prelude::*};
+
+mod benchmarking;
+mod mock;
+mod tests;
+
+pub use pallet::*;
+
+pub type PalletNameOf<T> = BoundedVec<u8, <T as Config>::MaxNameLen>;
+pub type ExtrinsicNameOf<T> = BoundedVec<u8, <T as Config>::MaxNameLen>;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(PhantomData<T>);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The overarching event type.
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		/// The only origin that can pause calls.
+		type PauseOrigin: EnsureOrigin<Self::Origin>;
+
+		/// The only origin that can un-pause calls.
+		type UnpauseOrigin: EnsureOrigin<Self::Origin>;
+
+		/// Pallets that are safe and can never be paused.
+		///
+		/// The tx-pause pallet is always assumed to be safe itself.
+		type UnpausablePallets: Contains<PalletNameOf<Self>>;
+
+		/// Maximum length for pallet- and extrinsic-names.
+		///
+		/// Too long names will not be truncated but handled like
+		/// [`Self::PauseTooLongNames`] specifies.
+		#[pallet::constant]
+		type MaxNameLen: Get<u32>;
+
+		/// Specifies if extrinsics and pallets with too long names should be treated as paused.
+		///
+		/// Setting this to `true` ensures that all calls that
+		/// are callable, are also pause-able.
+		/// Otherwise there could be a situation where a call
+		/// is callable but not pause-able, which would could be exploited.
+		#[pallet::constant]
+		type PauseTooLongNames: Get<bool>;
+
+		// Weight information for extrinsics in this pallet.
+		//type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The call is (already) paused.
+		IsPaused,
+
+		/// The call is (already) unpaused.
+		IsUnpaused,
+
+		/// The call is listed as safe and cannot be paused.
+		IsUnpausable,
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// This call got paused.
+		CallPaused(PalletNameOf<T>, ExtrinsicNameOf<T>),
+		/// This call got un-paused.
+		CallUnpaused(PalletNameOf<T>, ExtrinsicNameOf<T>),
+	}
+
+	/// The set of calls that are explicitly paused.
+	#[pallet::storage]
+	pub type PausedCalls<T: Config> =
+		StorageMap<_, Blake2_128Concat, (PalletNameOf<T>, ExtrinsicNameOf<T>), (), OptionQuery>;
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Pause a call.
+		///
+		/// Can only be called by [`Config::PauseOrigin`].
+		/// Emits an [`Event::CallPaused`] event on success.
+		#[pallet::weight(0)]
+		pub fn pause_call(
+			origin: OriginFor<T>,
+			pallet: PalletNameOf<T>,
+			extrinsic: ExtrinsicNameOf<T>,
```

Comment:
```suggestion
			call: CallNameOf<T>,
```

nit: Extrinsic is the broader term, in FRAME, we have calls and we are pausing calls, as the name of the transaction is suggesting.

Reply:
Ah okay. Then maybe `function`? I think a call is more general.

Reply:
Huh, yea it's not ideal. Then maybe `FunctionNameOf`? I thought that the combination of pallet+extrinsic makes a call, but pallet+function would make more sense.

Reply:
- `Dispatchable` is what we call function objects/functors (i.e. data which can be called) in Substrate in general.
- `Call` is the strict name of this data-structure in pallets (an enum).
- `Extrinsic` is a piece of data which can be included in a block and which leads to some action. Typical Substrate chains have extrinsics which contain a `Call` value.
- `Transaction` is an extrinsic which includes a signature so we know which account is authorising it.

Reply:
- `Dispatchable` is what we call function objects/functors (i.e. data which can be called) in Substrate in general.
- `Call` is the strict name of this data-structure in pallets (an enum).
- `Extrinsic` is a piece of data which can be included in a block and which leads to some action. Typical Substrate chains have extrinsics which contain a `Call` value.
- `Transaction` is an extrinsic which includes a signature so we know which account is authorising it.

Reply:
- `Dispatchable` is what we call function objects/functors (i.e. data which can be called) in Substrate in general.
- `Call` is the strict name of this data-structure in pallets (an enum).
- `Extrinsic` is a piece of data which can be included in a block and which leads to some action. Typical Substrate chains have extrinsics which contain a `Call` value.
- `Transaction` is an extrinsic which includes a signature so we know which account is authorising it.